docs/adr/0005-multitenancy-approach-for-marain.md

```diff
@@ -0,0 +1,263 @@
+ |
+ +-> Litware
+ |     +-> (Workflow storage configuration)
+ |     +-> (The Id of the WORKFLOW+Litware sub-tenant for the Workflow service to use)
+ |
+ +-> WORKFLOW
+ |     |
+ |     +-> WORKFLOW+Contoso
+ |     |     +-> (Operations storage configuration)
+ |     |     +-> (The Id of the OPERATIONS+WORKFLOW+Contoso sub-tenant for the Operations service to use)
+ |     |     +-> (FooBar storage configuration)
+ |     |
+ |     +-> WORKFLOW+Litware
+ |           +-> (Operations storage configuration)
+ |           +-> (The Id of the OPERATIONS+WORKFLOW+Litware sub-tenant for the Operations service to use)
+ |           +-> (FooBar storage configuration)
+ |
+ +-> OPERATIONS
+ |     |
+ |     +-> OPERATIONS+WORKFLOW+Contoso
+ |     |     +-> (FooBar storage configuration)
+ |     |
+ |     +-> OPERATIONS+WORKFLOW+Litware
+ |           +-> (FooBar storage configuration)
+ |
+ +-> FOOBAR
+```
+
+Since Litware is not licenced to use FooBar, the Litware Client Tenant does not hold any configuration for that service itself.
+
+Finally, we enroll Litware to use the Operations service. In this example, because Operations depends on FooBar, we need to create another sub-tenant of Operations to call FooBar with when Litware use Operations directly, and enroll this new subtenant with FooBar. This leaves us with the following:
+
+```
+Root tenant
+ |
+ +-> Contoso
+ |     +-> (Workflow storage configuration)
+ |     +-> (The Id of the WORKFLOW+Contoso sub-tenant for the Workflow service to use)
+ |     +-> (FooBar storage configuration)
+ |
+ +-> Litware
+ |     +-> (Workflow storage configuration)
+ |     +-> (The Id of the WORKFLOW+Litware sub-tenant for the Workflow service to use)
+ |     +-> (Operations storage configuration)
+ |     +-> (The Id of the OPERATIONS+Litware sub-tenant for the Operations service to use)
+ |
+ +-> WORKFLOW
+ |     |
+ |     +-> WORKFLOW+Contoso
+ |     |     +-> (Operations storage configuration)
+ |     |     +-> (The Id of the OPERATIONS+WORKFLOW+Contoso sub-tenant for the Operations service to use)
+ |     |     +-> (FooBar storage configuration)
+ |     |
+ |     +-> WORKFLOW+Litware
+ |           +-> (Operations storage configuration)
+ |           +-> (The Id of the OPERATIONS+WORKFLOW+Litware sub-tenant for the Operations service to use)
+ |           +-> (FooBar storage configuration)
+ |
+ +-> OPERATIONS
+ |     |
+ |     +-> OPERATIONS+WORKFLOW+Contoso
+ |     |     +-> (FooBar storage configuration)
+ |     |
+ |     +-> OPERATIONS+WORKFLOW+Litware
+ |     |     +-> (FooBar storage configuration)
+ |     |
+ |     +-> OPERATIONS+Litware
+ |           +-> (FooBar storage configuration)
+ |
+ +-> FOOBAR
+```
+
+## Consequences
+
+It is expected that these sub-tenants are created and configured as part of onboarding a new Tenant; as part of this process, the tenant will be enrolled for usage of each Marain service. The means by which this is expected to happen is covered in a separate ADR.
```

Comment:
You've combined together tenant onboarding and service enrollment here. I think it would be better to separate them out because we want to be able to take an existing tenant and enroll them into additional services. For example if, in 6 months, Litware see the error of their ways and ask if they could get direct access to Operations too, we need to be able to make that happen.

The significance of this is that not _all_ of the sub-tenant creation happens at onboarding time. Some of it happens at enrollment time. (If a tenant always enrolls into all the services it will ever use at onboarding time, this isn't a meaningful distinction. But it matters if we can enroll into additional services as time goes on.)

So I think we need to be clear. I think only the client Tenant is created at onboarding time. And then for each enrollment we add the necessary config to the client Tenant, and also create any necessary sub-tenants.

Reply:
Ah yes. That's not really what I intended to mean, but from a re-read it does come across that way.